### PR TITLE
Farewell, Legacy Python

### PR DESCRIPTION
**Pull request by jairideout**
~~**Not ready for merge!**~~

Removes Python 2 support as described in the [py3-only RFC](https://github.com/biocore/scikit-bio-rfcs/blob/master/active/002-py3-only.md). :guitar:

To-do:

- [x] install-time guard in setup.py, checklist.py, and doc/source/conf.py to prevent Python 2 users from installing/executing
- [x] ~~create issue re: front page disclaimer (RFC implementation step 3)~~ #1316
- [x] ~~create issue re: guide for Python 2 devs (RFC implementation step 4)~~ #1317
- [x] ~~create issue re: testing against latest Python dev version (RFC implementation step 8)~~ #1318

### CODE REVIEW

**Comment by wasade** [comment]
Woooooooo!!!!

**Comment by gregcaporaso** [comment]
Good idea re: check for python 2 @jairideout

File: checklist.py [hunk 1]
```diff
@@ -450,32 +448,5 @@ def _parse_file(self, fp, root):
         return skbio_imports
 
 
-class FluxCapacitorValidator(RepoValidator):
-    """Ensure that the __future__ statements are fluxing correctly"""
```
Comment (ebolyen) [inline checklist.py]: This was a valiant validator and will be missed. :cry:

File: skbio/io/format/tests/test_fasta.py [hunk 2]
```diff
@@ -714,8 +708,7 @@ def setUp(self):
         self.msa = TabularMSA(seqs)
 
         def empty_gen():
-            return
-            yield
+            yield from ()
```
Comment (ebolyen) [inline skbio/io/format/tests/test_fasta.py]: woah.

File: skbio/metadata/_repr.py [hunk 3]
```diff
@@ -84,13 +80,13 @@ def _format_metadata_key_value(self, key, value):
         key_fmt = self._format_key(key)
 
         supported_type = True
-        if isinstance(value, (six.text_type, six.binary_type)):
+        if isinstance(value, (str, bytes)):
```
Comment (ebolyen) [inline skbio/metadata/_repr.py]: I think `str` can be dropped now.

File: skbio/stats/_subsample.py [hunk 4]
```diff
@@ -6,23 +6,15 @@
...
 
 import numpy as np
 
-from skbio.util import EfficiencyWarning
-try:
-    from .__subsample import _subsample_counts_without_replacement
-except ImportError:
-    pass
 from skbio.util._decorator import experimental
+from .__subsample import _subsample_counts_without_replacement
```
Comment (ebolyen) [inline skbio/stats/_subsample.py]: :+1:

File: skbio/stats/distance/tests/test_base.py [hunk 5]
```diff
@@ -318,15 +314,14 @@ def test_plot_no_default(self):
...
         obs = dm._repr_png_()
-        self.assertIsInstance(obs, binary_type)
+        self.assertIsInstance(obs, bytes)
         self.assertTrue(len(obs) > 0)
 
     def test_repr_svg(self):
         obs = self.dm_1x1._repr_svg_()
-        # print_figure(format='svg') can return text or bytes depending on the
+        # print_figure(format='svg') can return str or bytes depending on the
         # version of IPython
```
Comment (ebolyen) [inline skbio/stats/distance/tests/test_base.py]: Is this still relevant for our current minimum IPython version?

File: skbio/tree/tests/test_majority_rule.py [hunk 6]
```diff
@@ -22,17 +20,26 @@
...
+            TreeNode.read(
+                io.StringIO("(A,(B,(E,(G,((F,I),(((J,H),D),C))))));")),
+            TreeNode.read(
+                io.StringIO("(A,(B,(E,((F,I),(G,((J,(H,D)),C))))));")),
+            TreeNode.read(
+                io.StringIO("(A,(B,(E,((F,I),(G,(((J,H),D),C))))));")),
+            TreeNode.read(
+                io.StringIO("(A,(B,(E,((G,(F,I)),((J,(H,D)),C)))));")),
+            TreeNode.read(
+                io.StringIO("(A,(B,(E,((G,(F,I)),(((J,H),D),C)))));"))]
```
Comment (ebolyen) [inline skbio/tree/tests/test_majority_rule.py]: Ouch. In this case, `map(TreeNode.read, [[newick-string], [newick-string], ...])` might have been easier than reindenting.

File: skbio/util/_misc.py [hunk 7]
```diff
@@ -29,7 +27,7 @@ def resolve_key(obj, key):
 
 
 def make_sentinel(name):
-    return type(name, (object, ), {
+    return type(name, tuple(), {
```
Comment (ebolyen) [inline skbio/util/_misc.py]: cool!

File: skbio/util/_misc.py [hunk 8]
```diff
@@ -37,22 +35,9 @@ def make_sentinel(name):
...
-            # Concept from http://stackoverflow.com/a/12627202/579416
-            keywords_start = -len(function_spec.defaults)
-            for key, default in zip(function_spec.args[keywords_start:],
-                                    function_spec.defaults):
-                if default is sentinel:
-                    keys.append(key)
-    return keys
+    params = inspect.signature(function).parameters
+    return [name for name, param in params.items()
+            if param.default is sentinel]
```
Comment (ebolyen) [inline skbio/util/_misc.py]: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1:

File: skbio/util/_misc.py [hunk 9]
```diff
@@ -78,15 +63,10 @@ def formatted_listing(self):
 
     def interpolate(self, obj, name):
         """Inject the formatted listing in the second blank line of `name`."""
-        # Py2/3 compatible way of calling getattr(obj, name).__func__
-        f = getattr(obj, name).__get__(None, type(None))
+        f = getattr(obj, name)
+        f2 = FunctionType(f.__code__, f.__globals__, name=f.__name__,
+                          argdefs=f.__defaults__, closure=f.__closure__)
```
Comment (ebolyen) [inline skbio/util/_misc.py]: I forgot about these horrible hacks.

File: skbio/util/tests/test_decorator.py [hunk 10]
```diff
@@ -225,11 +224,15 @@ def test_function_docstring(self):
...
-        self.assertEqual(inspect.getargspec(f), expected)
+
+        parameters = [
+            inspect.Parameter('x', inspect.Parameter.POSITIONAL_OR_KEYWORD),
+            inspect.Parameter('y', inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                              default=42)
+        ]
+        expected = inspect.Signature(parameters)
+
+        self.assertEqual(inspect.signature(f), expected)
```
Comment (ebolyen) [inline skbio/util/tests/test_decorator.py]: This is beautiful.

**Comment by ebolyen** [comment]
@jairideout Finished reviewing. This is awesome! I had two questions, but neither are very important. Let me know when you are ready for another round/merge.

File: skbio/util/_misc.py (hunk 7 above)
Comment (ebolyen) [inline skbio/util/_misc.py]: Can `tuple()` be simplified to `()`?

File: checklist.py (hunk 1 above)
Comment (jairideout) [inline checklist.py]: agree

File: skbio/metadata/_repr.py (hunk 3 above)
Comment (jairideout) [inline skbio/metadata/_repr.py]: Good catch, I'll need to rewrite this block

File: skbio/stats/distance/tests/test_base.py (hunk 5 above)
Comment (jairideout) [inline skbio/stats/distance/tests/test_base.py]: Probably not, will verify

File: skbio/tree/tests/test_majority_rule.py (hunk 6 above)
Comment (jairideout) [inline skbio/tree/tests/test_majority_rule.py]: twas simple with vi, you should try it sometime ;)

File: skbio/util/_misc.py (hunk 7 above)
Comment (jairideout) [inline skbio/util/_misc.py]: good catch

File: skbio/stats/distance/tests/test_base.py (hunk 5 above)
Comment (jairideout) [inline skbio/stats/distance/tests/test_base.py]: fixed

File: skbio/tree/tests/test_majority_rule.py (hunk 6 above)
Comment (anderspitman) [inline skbio/tree/tests/test_majority_rule.py]: that dup tho

**Comment by jairideout** [comment]
@ebolyen thanks for reviewing this monster! It's ready for another review/merge.

**Comment by ebolyen** [comment]
Thanks @jairideout!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by jairideout** [comment]
:beers:

**Pull request merged**